packages/transport-native/android/src/main/java/io/trezor/transport/bridges/USBBridge.java

```diff
@@ -0,0 +1,331 @@
+        request.queue(ByteBuffer.wrap(buffer), 64);
+        usbConnection.requestWait();
+      }
+      Log.i(TAG, "Writing done");
+    }
+
+    @Override
+    public byte[] rawRead() {
+      ByteBuffer data = null;// ByteBuffer.allocate(32768);
+      ByteBuffer buffer = ByteBuffer.allocate(64);
+      UsbRequest request = new UsbRequest();
+      request.initialize(usbConnection, readEndpoint);
+      int msg_size;
+      int invalidChunksCounter = 0;
+
+      // read first 64bytes
+      for (; ; ) {
+        request.queue(buffer, 64);
+        usbConnection.requestWait();
+        byte[] b = buffer.array();
+        Log.i(TAG, String.format("messageRead: Read chunk: %d bytes", b.length));
+
+        if (b.length < 9 || b[0] != (byte) '?' || b[1] != (byte) '#' || b[2] != (byte) '#') {
+          if (invalidChunksCounter++ > 5) {
+            Log.e(TAG, "THROW EXCEPTION");
+            throw new TrezorException("too many invalid chunks");
+          }
+          continue;
+        }
+        if (b[0] != (byte) '?' || b[1] != (byte) '#' || b[2] != (byte) '#')
+          continue;
+
+        msg_size = (((int) b[5] & 0xFF) << 24)
+          + (((int) b[6] & 0xFF) << 16)
+          + (((int) b[7] & 0xFF) << 8)
+          + ((int) b[8] & 0xFF);
+
+        data = ByteBuffer.allocate(msg_size + 1024);
+        data.put(b, 1, b.length - 1);
+        break;
+      }
+
+      invalidChunksCounter = 0;
+
+      // read the rest of the data
+      while (data.position() < msg_size) {
+        request.queue(buffer, 64);
+        usbConnection.requestWait();
+        byte[] b = buffer.array();
+        Log.i(TAG, String.format("messageRead: Read chunk (cont): %d bytes", b.length));
+        if (b[0] != (byte) '?') {
+          if (invalidChunksCounter++ > 5) {
+            Log.e(TAG, "THROW EXCEPTION");
+            throw new TrezorException("too many invalid chunks");
+          }
+          continue;
+        }
+
+        data.put(b, 1, b.length - 1);
+      }
+      int paddedLength = Utils.calculatePaddedLength(msg_size, 63);
+      Log.d(TAG, String.format("data size %d value %s", paddedLength, Utils.byteArrayToHexString(data.array())));
+
+      return Arrays.copyOfRange(data.array(), 0, paddedLength);
+    }
+
+    @Override
+    public void openConnection(Context context) throws TrezorException {
+      UsbManager usbManager = (UsbManager) context.getSystemService(Context.USB_SERVICE);
+
+      // use first interface
+      usbInterface = usbDevice.getInterface(0);
+      // try to find read/write endpoints
+      readEndpoint = null;
+      writeEndpoint = null;
+      for (int i = 0; i < usbInterface.getEndpointCount(); i++) {
+        UsbEndpoint ep = usbInterface.getEndpoint(i);
+        if (readEndpoint == null && ep.getType() == UsbConstants.USB_ENDPOINT_XFER_INT && ep.getAddress() == 0x81) { // number
+          // =
+          // 1
+          // ;
+          // dir
+          // =
+          // USB_DIR_IN
```

Comment:
nit: this comment was supposed to stay at one line probably? Is it prettier's work?

Reply:
yep, let's move the comment to line above so it will stay in one piece

Reply:
actually this is probably caused by some formatter from Android studio... I will fix it